inc/functions.php

```diff
@@ -569,49 +627,44 @@ function my_mail($to, $subject, $message, $from="", $charset="", $headers="", $k
 	global $mybb, $plugins;
 	static $mail;
 
-	// Does our object not exist? Create it
-	if(!is_object($mail))
-	{
-		require_once MYBB_ROOT."inc/class_mailhandler.php";
+	// Get our mail handler.
+	$mail = &get_mybb_mailhandler();
 
-		$plugins->run_hooks('my_mail_mailhandler_init', $mail);
-
-		if(!is_object($mail) || !($mail instanceof MailHandler))
+	$my_mail_parameters = compact("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email");
+	$my_mail_parameters = $plugins->run_hooks('my_mail_parameters', $my_mail_parameters);
+	if(is_array($my_mail_parameters))
+	{
+		foreach($my_mail_parameters as $key => $value)
 		{
-			if($mybb->settings['mail_handler'] == 'smtp')
+			// Using extract() here may not be safe, so we work it out manually.
+			if(in_array($key, array("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email")))
 			{
-				require_once MYBB_ROOT."inc/mailhandlers/smtp.php";
-				$mail = new SmtpMail();
-			}
-			else
-			{
-				require_once MYBB_ROOT."inc/mailhandlers/php.php";
-				$mail = new PhpMail();
+				$$key = $value;
 			}
 		}
 	}
+	$my_mail_parameters = compact("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email");
 
-	// Using SMTP based mail
-	if($mybb->settings['mail_handler'] == 'smtp')
+	// If MyBB's built-in SMTP mail handler is used, set the keep alive bit accordingly.
+	if(isset($mybb->settings['mail_handler']) && $mybb->settings['mail_handler'] == 'smtp' && $keep_alive == true && !empty($mail->keep_alive))
 	{
-		if($keep_alive == true)
+		require_once MYBB_ROOT . "inc/class_mailhandler.php";
+		require_once MYBB_ROOT . "inc/mailhandlers/smtp.php";
+		if($mail instanceof MailHandler && $mail instanceof SmtpMail)
 		{
 			$mail->keep_alive = true;
 		}
 	}
 
-	// Using PHP based mail()
-	else
-	{
-		if($mybb->settings['mail_parameters'] != '')
-		{
-			$mail->additional_parameters = $mybb->settings['mail_parameters'];
-		}
-	}
-
-	// Build and send
+	// Build the mail message.
 	$mail->build_message($to, $subject, $message, $from, $charset, $headers, $format, $message_text, $return_email);
-	return $mail->send();
+
+	// Send the mail via this hook, if any.
+	// The hook would return true/false if the mail is sent or not, otherwise anything else that is none boolean.
+	$sent = $plugins->run_hooks('my_mail_send', $my_mail_parameters);
+
+	// Or send the mail using the mail handler.
+	return is_bool($sent) === true ? $sent : $mail->send();
```

Comment:
If the return value is not boolean, it may carry some error message while trying to send the mail from within a plugin.
We are't considering that value and trying to send the mail again?

Reply:
We don't currently have mail message retrying built in, and on failure the sender should add a mail log.

Reply:
Current plugin system doesn't allow plugins to return values (#4230). I haven't got a whole solution here.

My idea was that, if there are several plugins hooked here `my_mail_send`, when executing its own code, the plugin should check if previous one has already send the mail, so that it'll do its job, sending the mail or something else.

I'm not sure I've made myself clear.

Reply:
I plan to compact the variable `$send` into `$my_mail_parameters`. Doing so would doubtless complicate the hook's implementations for plugins but make the hook itself cleaner, under the circumstance that types of a plugin's return are limited, especially we might want to completely remove the ability to return from plugins.